doc/crdt-details.md

```diff
@@ -0,0 +1,1054 @@
+
+```
+()  Revision: before any revisions are made
+    Engine:   text="" tombstones="" deletes_from_union={}
+    Concepts: union="" back_computed_deletions_from_6_union={0,1,2,3,4,5}
+(1) Revision: inserts={0, 1, 2} deletes={}
+    Engine:   text="123" tombstones="" deletes_from_union={}
+    Concepts: union="123" back_computed_deletions_from_6_union={0,1,2}
+(2) Revision: inserts={} deletes={0, 1, 2}
+    Engine:   text="" tombstones="123" deletes_from_union={0,1,2}
+    Concepts: union="123" back_computed_deletions_from_6_union={0,1,2,3,4,5}
+(3) Revision: inserts={0} deletes={}
+    Engine:   text="a" tombstones="123" deletes_from_union={1,2,3}
+    Concepts: union="a123" back_computed_deletions_from_6_union={1,2,3,4,5}
+(4) Revision: inserts={1} deletes={}
+    Engine:   text="ab" tombstones="123" deletes_from_union={2,3,4}
+    Concepts: union="ab123" back_computed_deletions_from_6_union={1,3,4,5}
+(5) Revision: inserts={1} deletes={}
+    Engine:   text="axb" tombstones="123" deletes_from_union={3,4,5}
+    Concepts: union="axb123" back_computed_deletions_from_6_union={3,4,5}
+(6) Revision: inserts={} deletes={1}
+    Engine:   text="ab" tombstones="x123" deletes_from_union={1,3,4,5}
+    Concepts: union="axb123"
+```
+
+## Operations
+
+Now that you know what we have to work with, let's go over the operations that `Engine` supports. Each of these operations relies on a bunch of different shared helpers, as we go from the simplest operations to the most complex, we'll gradually build up the set of helpers we use.
+
+After describing how each operation or helper works there'll often be a code block with the actual function, it isn't necessary to understand the code and it may have complications that aren't mentioned, so feel free to skip them, they're there if you want to confirm or enhance your understanding.
+
+### Subset helpers
+
+`Subset` has a number of operations that produce new `Subset`s, these form the core of most work the CRDT operations do. It's better to explain them together near the start since they're used everywhere:
+
+#### Subset::union
+
+Takes two `Subset`s and produces a new `Subset` of the same string where each character has the sum of the counts it has in each input. When treating a `Subset` as a normal set, this is just the union.
+
+#### Subset::transform_expand
+
+`Subset::transform_expand` takes a `Subset` and another "transform" `Subset` and transforms the first `Subset` through the coordinate transform represented by the "transform". Now what does this mean:
+
+`Revision`s are never modified even when the union string expands with newly inserted characters. We can deal with this by treating `Subset`s of inserted characters as coordinate transforms. The only difference to the union string is the inserted characters, so if we can modify the coordinates of a `Subset` from being based on one union string to another, we can work with edits from multiple `Revision`s together.
```

Comment:
>`Revision`s are never modified even when the union string expands with newly inserted characters.

This could be clearer. Like: "`Revision`s are never modified, and their `Edit`s always operate on the union string that exists when they are created. For operate on multiple `Revision`s, have to be able to map coordinates in one's union string to coordinates in another's. We do this by treating [...]`


And a question:

When we use a subset as a transform is it really a subset, or is this just an incidental homomorphism, where the `Subset` type just happens to be share a representation with some `SubsetTransform` type, and so we just overload it and get on with things? This operation felt like a strange use of a `Subset`, and it took me a few read-throughs to get a sense of what is going on.